### DISCUSSION TITLE
Issues with accessing descriptors that return a callable from class instances

### DISCUSSION THREAD

**Discussion by StockerMC** [microsoft/pyright#2508] [Q&A]
Consider the following code (in a pyi file):
```python
from _typeshed import Self
from functools import _CacheInfo  # type: ignore
from typing import Generic, overload, Callable, TypeVar, Coroutine, Any, Protocol, Hashable
from typing_extensions import ParamSpec, Concatenate

_T = TypeVar("_T", contravariant=True)
_O = TypeVar("_O")
_P = ParamSpec("_P")
_Coro = Coroutine[Any, Any, _T]

__version__: str

class _callable_alru_cache(Protocol[_P, _T]):
    _origin: Callable[_P, _Coro[_T]]
    closed: bool
    async def __call__(self, *args: _P.args, **kwargs: _P.kwargs) -> _T: ...
    def cache_info(self) -> _CacheInfo: ...
    def cache_clear(self) -> None: ...
    def invalidate(self, *args: Hashable, **kwargs: Hashable) -> bool: ...
    def close(self, *, cancel: bool = ..., return_exceptions: bool = ...) -> _Coro[list[_T]]: ...
    def open(self) -> None: ...

class _wrapped_alru_cache(_callable_alru_cache[_P, _T], Generic[_O, _P, _T]):
    @overload
    def __get__(self: Self, instance: None, _) -> Self: ...
    @overload
    def __get__(
        self,
        instance: _O,
        owner: type[_O],
    ) -> _callable_alru_cache[Concatenate[_O, _P], _T]: ...

@overload
def alru_cache(
    fn: None = ...,
    maxsize: int | None = ...,
    typed: bool = ...,
    *,
    cache_exceptions: bool = ...,
) -> Callable[[Callable[_P, _Coro[_T]]], _wrapped_alru_cache[Any, _P, _T]]: ...

@overload
def alru_cache(
    fn: Callable[_P, _Coro[_T]],
    maxsize: int | None = ...,
    typed: bool = ...,
    *,
    cache_exceptions: bool = ...,
) -> _wrapped_alru_cache[Any, _P, _T]: ...
```
With the following example:
![image](https://user-images.githubusercontent.com/44980366/139312265-72458db4-3bd2-4bb9-ac05-0c898ef85e45.png)

When a method decorated with `alru_cache` is accessed from an instance of the class it is in, pyright will say that it takes no positional or keyword arguments. If the method is instead accessed from the class itself, pyright will recognize its parameters. If a function is decorated outside of a class, pyright will recognize its parameters as well.

The type of `await self.foo()` should also be `int`, not `_T@callable_alru_cache`, though that may be directly related to the issue with the parameters.

How would I correctly type this?

**Comment by erictraut** [comment]
You're using `ParamSpec` and `Concatenate` in creative ways that I haven't seen before! :)

You've uncovered at least one bug in pyright, which I've fixed. It has to do with the explicit specialization of a generic class using `Concatenate` within one of the type argument expressions. The return type `_callable_alru_cache[Concatenate[_O, _P], _T]` was being incorrectly handled. I've addressed that bug.

I think there are still bugs in your code though.

Let's take a look at a subset of the code from your screen shot above. 

```python
class A:
    @alru_cache
    async def foo(self, a: int) -> int:
        ...

    async def other(self) -> None:
        reveal_type(self.foo)
        reveal_type(await self.foo(2))
```

If there was no `__get__` method, the type of the expression `self.foo` would be `_wrapped_alru_cache[Any, (self: A, a: int), int]`.

Once the descriptor is applied, the type of `self.foo` becomes `_callable_alru_cache[(__p0: Any, self: A, a: int), int]`.

That means the specialized signature of the `_callable_alru_cache.__call__` method becomes `async def __call__(self, __p0: Any, self: A, a: int) -> int: ...```.

When you attempt to invoke this with a single argument expression `2`, pyright should complain that you are missing arguments for parameters `self` and `a`. With my bug fix in place, this is the error you will see.

![Screen Shot 2021-10-28 at 4 27 02 PM](https://user-images.githubusercontent.com/7040122/139349785-bc972283-ebe7-4400-8dee-f9544e0e02f0.png)

**Reply by StockerMC** [reply]
Thanks! What I was trying to do with this code was to get pyright to not complain about a missing argument for self when the method was accessed from an instance. The code you posted with the actual implementation (the actual module was installed from `git+https://github.com/aio-libs/async-lru`) would run without error at runtime (without the `reveal_type`s of course). Is there a possible way to indicate that `self` doesn't need to be passed when accessed from an instance? I was under the impression that my use of `Concatenate` with `__get__` handled that, but it doesn't seem to handle it how I intended it to.

A small example using the actual library:
- Install `git+https://github.com/aio-libs/async-lru` with `pip`
```python
from async_lru import alru_cache
import asyncio

@alru_cache
async def not_in_class(a: int) -> int:
    return a + 1

class A:
    @alru_cache
    async def in_class(self, a: int) -> int:
        return a + 1

async def main():
    print(await not_in_class(1)) # 2
    print(await A().in_class(2)) # 3

asyncio.run(main())
```

**Reply by StockerMC** [reply]
@erictraut In addition to the above problem, the parameters of the `_origin` attribute of the `_callable_alru_cache` protocol aren't being set on instances.

![image](https://user-images.githubusercontent.com/44980366/139739483-00238848-1d06-47ba-a411-1d4519ac4cb6.png)

Apologies, I don't believe pylance is on pyright 1.1.184 yet, though the above problem is still an issue for me.

The pyright CLI shows:
```python
info: Type of "A.in_class" is "_wrapped_alru_cache[(self: A, a: int), int]"
info: Type of "A().in_class" is "_callable_alru_cache[(A, self: Self@A, a: int), int]"   
info: Type of "A.in_class._origin" is "(self: Self@A, a: int)"
info: Type of "A().in_class._origin" is "(_O@__get__, self: Self@A, a: int)"
error: Arguments missing for parameters "self", "a" (reportGeneralTypeIssues)
error: Arguments missing for parameters "__p0", "self", "a" (reportGeneralTypeIssues)
```

It is requiring `__p0` as an argument, and is still requiring self to be passed even though it is accessed from an instance.

**Reply by erictraut** [reply]
Could you post an updated code sample that you think should type check without errors? I presume that you've addressed the issues in your previous sample that I mentioned above.

**Reply by StockerMC** [reply]
I don't fully understand the part about self being required if it's accessed from an instance (from the screenshot you showed), though this code (that I would expect to type check without errors) may demonstrate what I mean more clearly.
```python
class A:
    @alru_cache
    async def in_class(self, a: int) -> int:
        return a + 1

async def main():
    await A().in_class(1)
    await A().in_class._origin(1)
```

Both function calls produce `Arguments missing for parameters "self", "a"`
Actually, the second function call should in fact produce that error (unlike the first).

**Reply by erictraut** [reply]
This code snippet is incomplete. Is your assertion that it should work with the unmodified example at the top of this discussion thread? That code has problems that must be addressed.

The sample defines a decorator `alru_cache`. If I extract the applicable overload and remove the irrelevant parameters, I'm left with the following signature:

```python
def alru_cache(fn: Callable[_P, _Coro[_T]]) -> _wrapped_alru_cache[Any, _P, _T]: ...
```

When you apply this decorator to the method `async def in_class(self, a: int) -> int`, you're left with the type `_wrapped_alru_cache[Any,(self: A, a: int), int]`.

However, the `_wrapped_alru_cache` implements a descriptor protocol. Once the descriptor's `__get__` method is applied, we're left with the type `_callable_alru_cache[(__p0: Any, self: A, a: int), int]`.

The `_callable_alru_cache` class implements a `__call__` method. When we replace the type variables with the corresponding specialized types, we get the following signature: `async def __call__(self, __p0: Any, self: A, a: int) -> int`. This signature requires three arguments, but you are providing only one argument in the expression `A().in_class(1)`.

**Reply by StockerMC** [reply]
@erictraut Here is my updated code, in which I removed the use of Concatenate, which seemed to partially help:
```python
from _typeshed import Self
from functools import _CacheInfo  # type: ignore
from typing import Generic, overload, Callable, TypeVar, Coroutine, Any, Protocol, Hashable
from typing_extensions import ParamSpec

_T = TypeVar("_T", contravariant=True)
_O = TypeVar("_O")
_P = ParamSpec("_P")
_Coro = Coroutine[Any, Any, _T]

class _callable_alru_cache(Protocol[_P, _T]):
    _origin: Callable[_P, _Coro[_T]]
    closed: bool
    async def __call__(self, *args: _P.args, **kwargs: _P.kwargs) -> _T: ...
    def cache_info(self) -> _CacheInfo: ...
    def cache_clear(self) -> None: ...
    def invalidate(self, *args: Hashable, **kwargs: Hashable) -> bool: ...
    def close(self, *, cancel: bool = ..., return_exceptions: bool = ...) -> _Coro[list[_T]]: ...
    def open(self) -> None: ...

class _wrapped_alru_cache(_callable_alru_cache[_P, _T], Generic[_O, _P, _T]):
    @overload
    def __get__(self: Self, instance: None, _) -> Self: ...
    @overload
    def __get__(
        self,
        instance: _O,
        owner: type[_O],
    ) -> _callable_alru_cache[_P, _T]: ...

@overload
def alru_cache(
    fn: None = ...,
    maxsize: int | None = ...,
    typed: bool = ...,
    *,
    cache_exceptions: bool = ...,
) -> Callable[[Callable[_P, _Coro[_T]]], _wrapped_alru_cache[Any, _P, _T]]: ...

@overload
def alru_cache(
    fn: Callable[_P, _Coro[_T]],
    maxsize: int | None = ...,
    typed: bool = ...,
    *,
    cache_exceptions: bool = ...,
) -> _wrapped_alru_cache[Any, _P, _T]: ...

class A:
    @alru_cache
    async def in_class(self, a: int) -> int:
        return a + 1

    def meth(self, a: int) -> int: ...

async def main():
    reveal_type(A.in_class)
    reveal_type(A().in_class)
    await A().in_class(1) # Incorrect error - we passed a, and self was implicitly passed before a, therefore there would not be an error at runtime
    await A.in_class()  # Correct error - self and a weren't passed
    await A().in_class._origin(1)  # Correct error - we didn't pass an argument for a
```
![image](https://user-images.githubusercontent.com/44980366/139777917-02f1e291-32cd-41ca-8512-39a28d380e29.png)

I apologize if I'm not understanding something obvious from what you have told me. I am essentially trying to indicate in the method's signature that self doesn't need to be passed, because it is implicitly passed when accessed from an instance of the class it is in. In the provided example, I would like the parameters in the signature of `A().in_class` to be the same as if it wasn't decorated with the `alru_cache` decorator and was a regular method (`(a: int) -> Coroutine[Any, Any, int]` = `_callable_alru_cache[(a: int), int]`).

**Reply by erictraut** [reply]
I think I understand the disconnect. When `Concatenate` is used in an input parameter, the concatenated types are matched against the parameters of the callable type. This removes the first n parameters (where n is the number of types in the Concatenate expression), and only the remaining parameters in the signature are captured by the ParamSpec. Conversely, when `Concatenate` is used in a return type, it prepends these parameters to the signature captured by the ParamSpec. For example, if you want to replace the first parameter (which must be an `int`) with a different parameter (which is a `str`), that would look like this:

`def func(callback: Callable[Concatenate[int, P], R]) -> Callable[Concatenate[str, P], R]: ...`

In your original example, you were using `Concatenate` in a return type and expecting it to work as it would in the case of an input parameter annotation. That explains why `Concatenate` was not working the way you had intended.

Out of curiosity, how does your implementation bind the decorated method to the object instance? Presumably that's done in the `_wrapped_alru_cache.__get__` method, since it has access to the instance? If my supposition is correct, then the last line in your sample should not generate an error at runtime because `_origin` will already be bound to an instance of `A` at the time it is called.

Modeling the behavior of your code with ParamSpec and Concatenate is challenging. The best solution I can offer is to strip off the `self` parameter immediately within the `alru_cache` decorator. You would do this by annotating the `fn` parameter as follows: `Callable[Concatenate[Any, _P], _Coro[_T]]`. I think this will (mostly) give you the behavior you're looking for. It's not a very satisfying solution though because it doesn't correctly model how the runtime code works. At runtime, the `self` is not eliminated by `alru_cache` decorator but is instead eliminated once the method is bound to the object instance, which I presume occurs in `_wrapped_alru_cache.__get__`. So the approach I'm suggesting is a bit of a lie to appease the type checker.

**Reply by StockerMC** [reply]
@erictraut I appreciate you taking the time to help me with this.

> In your original example, you were using `Concatenate` in a return type and expecting it to work as it would in the case of an input parameter annotation. That explains why `Concatenate` was not working the way you had intended.

Thanks for the explanation.

> Out of curiosity, how does your implementation bind the decorated method to the object instance? Presumably that's done in the `_wrapped_alru_cache.__get__` method, since it has access to the instance? If my supposition is correct, then the last line in your sample should not generate an error at runtime because `_origin` will already be bound to an instance of `A` at the time it is called.

[This decorator](https://github.com/aio-libs/async-lru/blob/17a5d207f603c0c142f01a69fbf6f487b3fef5c4/async_lru.py#L119-L193) returns the callable itself, and not a wrapper class as I have done with the `_wrapped/callable_alru_cache` The instance gets passed as the first parameter because the method is being called from the instance. A small example is:
```python
import functools

def alru_cache(fn):
    @functools.wraps(fn)
    async def wrapped(*args, **kwargs):
        print(args, kwargs)
        return await fn(*args, **kwargs)
    return wrapped

class A:
    @alru_cache
    async def foo(self, a):
        ...

a = A()
await a.foo(3)
(<__main__.A object at 0x000001A2F3385360>, 3) {}
```
The `_origin` attribute is the function being decorated, and the decorator does not have access to the instance at the time, so the function itself is returned without the self parameter (`__self__` binded to it).

I'm not sure whether something that would do the opposite of `Concatenate` (not exactly, though) would help here, possibly to remove parameters from the `ParamSpec`, similar to how you mentioned `Concatenate` acted by removing the first n parameters. Although this doesn't exist currently, the example below would have `Deconcatenate` take a `ParamSpec` as it's last type argument, and any number of types preceding the `ParamSpec`, and would remove n parameters with the matching types from left to right.
```python
# format: Deconcatenate[...] == the signature of the "deconcatenated" param spec

# given _P == (self: A, a: int)
Deconcatenate[A, _P] == "(a: int)"
Deconcatenate[Any, _P] == "(a: int)"
# given _P == (self: A, a: int, b: str, c: bytes)
Deconcatenate[A, int, _P] == "(b: str, c: bytes)"

# and in my use case
class _wrapped_alru_cache(_callable_alru_cache[_P, _T], Generic[_O, _P, _T]):
    @overload
    def __get__(self: Self, instance: None, _) -> Self: ...
    @overload
    def __get__(
        self,
        instance: _O,
        owner: type[_O],
    ) -> _callable_alru_cache[Deconcatenate[_O, _P], _T]: ...

def alru_cache(fn: Callable[_P, _Coro[_T]]) -> _wrapped_alru_cache[Any, _P, _T]: ...
```

> The best solution I can offer is to strip off the self parameter immediately within the `alru_cache` decorator. You would do this by annotating the `fn` parameter as follows: `Callable[Concatenate[Any, _P], _Coro[_T]]`

Unfortunately this wouldn't work for functions that are not in a class.
